mmedit/models/editors/guided_diffusion/modules.py

```diff
@@ -0,0 +1,576 @@
+import math
+from copy import deepcopy
+from functools import partial
+
+import mmengine
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn.bricks import build_norm_layer
+from mmengine.model import constant_init
+from mmengine.utils.dl_utils import TORCH_VERSION
+from mmengine.utils.version_utils import digit_version
+
+from mmedit.registry import MODELS, MODULES
+
+
+class EmbedSequential(nn.Sequential):
+    """A sequential module that passes timestep embeddings to the children that
+    support it as an extra input.
+
+    Modified from
+    https://github.com/openai/improved-diffusion/blob/main/improved_diffusion/unet.py#L35
+    """
+
+    def forward(self, x, y):
+        for layer in self:
+            if isinstance(layer, DenoisingResBlock):
+                x = layer(x, y)
+            else:
+                x = layer(x)
+        return x
+
+
+@MODELS.register_module('GN32')
+class GroupNorm32(nn.GroupNorm):
+
+    def __init__(self, num_channels, num_groups=32, **kwargs):
+        super().__init__(num_groups, num_channels, **kwargs)
+
+    def forward(self, x):
+        return super().forward(x.float()).type(x.dtype)
+
+
+def convert_module_to_f16(layer):
+    """Convert primitive modules to float16."""
+    if isinstance(layer, (nn.Conv1d, nn.Conv2d, nn.Conv3d)):
+        layer.weight.data = layer.weight.data.half()
+        if layer.bias is not None:
+            layer.bias.data = layer.bias.data.half()
+
+
+def convert_module_to_f32(layer):
+    """Convert primitive modules to float32, undoing
+    convert_module_to_f16()."""
+    if isinstance(layer, (nn.Conv1d, nn.Conv2d, nn.Conv3d)):
+        layer.weight.data = layer.weight.data.float()
+        if layer.bias is not None:
+            layer.bias.data = layer.bias.data.float()
+
+
+@MODELS.register_module()
+class SiLU(nn.Module):
+    r"""Applies the Sigmoid Linear Unit (SiLU) function, element-wise.
+    The SiLU function is also known as the swish function.
+    Args:
+        input (bool, optional): Use inplace operation or not.
+            Defaults to `False`.
+    """
+
+    def __init__(self, inplace=False):
+        super().__init__()
+        if digit_version(TORCH_VERSION) <= digit_version('1.6.0') and inplace:
+            mmengine.print_log(
+                'Inplace version of \'SiLU\' is not supported for '
+                f'torch < 1.6.0, found \'{torch.version}\'.')
+        self.inplace = inplace
+
+    def forward(self, x):
+        """Forward function for SiLU.
+        Args:
+            x (torch.Tensor): Input tensor.
+
+        Returns:
+            torch.Tensor: Tensor after activation.
+        """
+
+        if digit_version(TORCH_VERSION) <= digit_version('1.6.0'):
+            return x * torch.sigmoid(x)
+
+        return F.silu(x, inplace=self.inplace)
+
+
+@MODULES.register_module()
+class MultiHeadAttention(nn.Module):
+    """An attention block allows spatial position to attend to each other.
```

Comment:
The name of this class looks similar to `MultiHeadAttentionBlock`, and this may make users confused.

Reply:
This class is removed now.